### PR TITLE
New version to support Azure Databricks 11.0 and upper

### PR DESCRIPTION
**Pull request by florent-brosse**
This version should be tested.
The goal is to share our work and see if you want to contribute/test.

The logger is now more dynamic, you can log any spark configuration you want and change the layout.
We had to remove the unified spark listener because the former implementations uses some internal spark class.
We create different listeners for events, query, streaming.

fixes: #187 
fixes: #204

### CODE REVIEW

**Comment by florent-brosse** [comment]
@microsoft-github-policy-service agree

**Review by pythr-prog — COMMENTED** [review]
Hi, I have tried the spark-monitoring library for Runtime 11 and linked it to my Databricks workspace and connected it to Log Analytics workspace. I have successfully attached the init script to my cluster. However, when I run jobs on this cluster, I do not see any logs of the Databricks jobs in Log Analytics. Does anyone have the same problem and has it resolved?

**Comment by HannaSasina** [comment]
I was to follow through all instructions and create a cluster. Built with Docker. However when I initialize the logger with below code it does not run (Waiting to run for ever)
Logger = spark._jvm.org.apache.log4j.Logger
mylogger = Logger.getLogger(...)

The sample job provided does not run either with status Pending for ever.

Please advise. Thanks!

**Comment by Paulsubs14** [comment]
Can i ask when this will be merge to the master branch?

**Comment by lukas1597** [comment]
I'm having the same issue as @DataNerdick. The cluster starts fine, but when I try to run any command (and not just the logger, i.e. dbutils), the status is stuck on "Waiting to run...".

**Comment by renepajta** [comment]
Hi all, 

what is currently blocking the PR to get merged into main? I have quite few customers waiting for this feature.

**Comment by florent-brosse** [comment]
Hi @lukas1597 and @DataNerdick could you please retry your test? Indeed there was a regression I've fixed.

**Comment by DataNerdick** [comment]
> Hi @lukas1597 and @DataNerdick could you please retry your test? Indeed there was a regression I've fixed.

         I no longer have a "Waiting to run" problem but logs are not following through to Log Analytics.

**Comment by lukas1597** [comment]
@DataNerdick Thanks, the issue is fixed for me at least. I can see logs, events and metrics are being populated in Log Analytics. Tested both sample batch and streaming jobs. I see regex filtering for logs and events is absent, might be good to add it in the future.

**Comment by DataNerdick** [comment]
> @florent-brosse Thanks, the issue is fixed for me at least. I can see logs, events and metrics are being populated in Log Analytics. Tested both sample batch and streaming jobs. I see regex filtering for logs and events is absent, might be good to add it in the future.

I wonder why it is still not working for me. Which profile have you built? How are you building it? Have you done any modifications to the code? Neither Scala 2.12_Spark 3.3.1 nor scala-2.12_spark-3.3.0 profiles send Logs for me

**Comment by lukas1597** [comment]
> > @florent-brosse Thanks, the issue is fixed for me at least. I can see logs, events and metrics are being populated in Log Analytics. Tested both sample batch and streaming jobs. I see regex filtering for logs and events is absent, might be good to add it in the future.
> 
> I wonder why it is still not working for me. Which profile have you built? How are you building it? Have you done any modifications to the code? Neither Scala 2.12_Spark 3.3.1 nor scala-2.12_spark-3.3.0 profiles send Logs for me

I'm using maven to build the jar. But there's no profiles in the pom file (they have been removed in this PR), there's only one configuration for spark 3.3.1, scala 2.12, so I didn't change anything in the library. Though I had to adjust the streaming job code to include the streaming listener.

**Comment by DataNerdick** [comment]
> > > @florent-brosse Thanks, the issue is fixed for me at least. I can see logs, events and metrics are being populated in Log Analytics. Tested both sample batch and streaming jobs. I see regex filtering for logs and events is absent, might be good to add it in the future.
> > 
> > 
> > I wonder why it is still not working for me. Which profile have you built? How are you building it? Have you done any modifications to the code? Neither Scala 2.12_Spark 3.3.1 nor scala-2.12_spark-3.3.0 profiles send Logs for me
> 
> I'm using maven to build the jar. But there's no profiles in the pom file (they have been removed in this PR), there's only one configuration for spark 3.3.1, scala 2.12, so I didn't change anything in the library. Though I had to adjust the streaming job code to include the streaming listener.

Thanks for your reply. Scala version can be changed in pom to 3.3.0, so I tried building both. With 3.3.1 are you choosing 12.0 Runtime on the cluster? Would you mind sharing the code you adjusted?
image

**Comment by gfree76** [comment]
This version is working for my team. I'd like to see it checked in soon so we can use it in our production. Who needs to review before it can be merged?

Also - does anyone know where i can find a guide to what fields are available for the sparkLayout.json? I'd like to include clustername and clusterId in our logging. Thanks!

**Review by scottmacchia-msft — APPROVED** [review]

**Comment by koichino** [comment]
Hi, this update will support DBR 12.2 LTS (Scala 2.12, Spark 3.3.2)? Users would like to use LTS version.

**Comment by braischao** [comment]
> Hi, this update will support DBR 12.2 LTS (Scala 2.12, Spark 3.3.2)? Users would like to use LTS version.

I have just tested it in DBR 12.2 LTS doing a few modifications in the Spark configuration and also uploading the Log4j 2.18.0 template. So far it seems to be working fine and logs are arriving properly.

**Comment by gfree76** [comment]
@florent-brosse - could you please check this in?

File: README.md [hunk 1]
```diff
@@ -91,7 +93,7 @@ docker run -it --rm -v %cd%:/spark-monitoring -v "%USERPROFILE%/.m2":/root/.m2 -
 
 1. Import the Maven project object model file, _pom.xml_, located in the **/** folder into your project. This will import the project
 
-1. Activate a **single** Maven profile that corresponds to the versions of the Scala/Spark combination that is being used. By default, the Scala 2.12 and Spark 3.3.1 profile is active.
+1. Activate a **single** Maven profile that corresponds to the versions of the DBR version combination that is being used. By default, the DBR version 12.2LTS profile is active.
```
Comment (gfree76) [inline README.md:96]: nitpick, non blocking: should be either "Scala/Spark combination" or "DBR version", but not "DBR version combination"

File: docs/filtering.md [hunk 2]
```diff
@@ -8,13 +8,27 @@ The Spark Monitoring Library can generate large volumes of logging and metrics d
 
 > Note: The REGEX value(s) should be surrounded by double quotes as noted in the examples so that the characters in the regular expression(s) are not interpretted by the shell.
 
-## Limiting events in SparkListenerEvent_CL
+## Adding or removing fiels in SparkEvents_CL
```
Comment (gfree76) [inline docs/filtering.md:11]: nitpick, non blocking: should be "Adding or removing fiel**d**s"

File: docs/filtering.md [hunk 3]
```diff
@@ -8,13 +8,27 @@ The Spark Monitoring Library can generate large volumes of logging and metrics d
...
 
-You can uncomment and edit the `LA_SPARKLISTENEREVENT_REGEX` environment variable that is included in [spark-monitoring.sh](../src/spark-listeners/scripts/spark-monitoring.sh) to limit the logging to only include events where Event_s matches the regex.
+In SparkLayout.json, you can specify your own template that would be use for logging.
```
Comment (gfree76) [inline docs/filtering.md:13]: nitpick, nonblocking: "In SparkLayout.json, you can specify your own template for logging."

File: docs/filtering.md [hunk 4]
```diff
@@ -8,13 +8,27 @@ The Spark Monitoring Library can generate large volumes of logging and metrics d
 
 > Note: The REGEX value(s) should be surrounded by double quotes as noted in the examples so that the characters in the regular expression(s) are not interpretted by the shell.
 
-## Limiting events in SparkListenerEvent_CL
+## Adding or removing fiels in SparkEvents_CL
 
-You can uncomment and edit the `LA_SPARKLISTENEREVENT_REGEX` environment variable that is included in [spark-monitoring.sh](../src/spark-listeners/scripts/spark-monitoring.sh) to limit the logging to only include events where Event_s matches the regex.
+In SparkLayout.json, you can specify your own template that would be use for logging.  
+The syntax used is the same as in log4j2 JSON Template: https://logging.apache.org/log4j/2.x/manual/json-template-layout.html.  
+As provided in SparkLayout.json, you can also add your own Spark variables, the full list can be found in the Environment tab on the SparkUi
```
Comment (gfree76) [inline docs/filtering.md:15]: nitpick, non blocking: "As provided in SparkLayout.json, you can also add your own Spark variables. The full list of available variables can be found in the Environment tab in the SparkUI."

File: build.sh [hunk 5]
```diff
@@ -7,5 +7,5 @@ echo "127.0.0.1   $(hostname)" >> /etc/hosts
 MAVEN_PROFILES=( "scala-2.12_spark-3.0.1" "scala-2.12_spark-3.1.2" "scala-2.12_spark-3.2.1" )
```
Comment (srvmsr) [inline build.sh:7]: You may need to update these values to match the profile values pom.xml

**Comment by gfree76** [comment]
Any reason this isn't being merged?

File: docs/filtering.md [hunk 6]
```diff
@@ -25,23 +39,6 @@ SparkListenerEvent_CL
...
-* SparkListenerStageSubmitted
-* SparkListenerTaskStart
-* SparkListenerTaskEnd
-* org.apache.spark.sql.streaming.StreamingQueryListener$QueryProgressEvent
-* SparkListenerStageCompleted
-* SparkListenerJobEnd
-
 ## Limiting Metrics in SparkMetric_CL
 
 You can uncomment and edit the `LA_SPARKMETRIC_REGEX` environment variable that is included in [spark-monitoring.sh](../src/spark-listeners/scripts/spark-monitoring.sh) to limit the logging to only include events where name_s matches the regex.
```
Comment (ChielvanGriethuysen) [inline docs/filtering.md:44]: Is this functionality still supported? See that the line LA_SPARKMETRIC_REGEX has disappeared from the spark-monitoring.sh file. Is there a new way to limit the amount of spark metrics produced?

**Comment by fanxinshr** [comment]
When I tried to use this new version, I found the file spark-sample-job-1.0.0.jar instead of spark-jobs-1.0-SNAPSHOT.jar in target folder after building.
Is this expected result ?

image

**Pull request merged**